File: static/cpp/top_k.cc
// Copyright (c) 2015 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <queue>
#include <random>
#include <sstream>
#include <string>
#include <vector>

using std::default_random_engine;
using std::equal;
using std::function;
using std::greater;
using std::istringstream;
using std::priority_queue;
using std::random_device;
using std::stoi;
using std::string;
using std::to_string;
using std::uniform_int_distribution;
using std::vector;
using std::cout;
using std::endl;

// @include
vector<string> TopK(int k, istringstream* stream) {
  priority_queue<string, vector<string>, function<bool(string, string)>>
  min_heap(
      [](const string& a, const string& b) { return a.size() >= b.size(); });
  string next_string;
  while (*stream >> next_string) {
    min_heap.emplace(next_string);
    if (min_heap.size() > k) {
      // Remove the shortest string. Note that the comparison function above
      // will order the strings by length.
      min_heap.pop();
    }
  }
  vector<string> result;
  while (!min_heap.empty()) {
    result.emplace_back(min_heap.top());
    min_heap.pop();
  }
  return result;
}
// @exclude

int main(int argc, char* argv[]) {
  default_random_engine gen((random_device())());
  for (int times = 0; times < 1000; ++times) {
    int num;
    if (argc == 2) {
      num = stoi(argv[1]);
    } else {
      uniform_int_distribution<int> num_dis(1, 10000);
      num = num_dis(gen);
    }
    vector<string> A;
    uniform_int_distribution<int> dis(1, 10000000);
    for (int i = 0; i < num; ++i) {
      A.emplace_back(to_string(dis(gen)));
    }
    uniform_int_distribution<int> kDis(1, num);
    int k = kDis(gen);
    string input_stream;
    for (const auto& s : A) {
      input_stream += s + ' ';
    }
    cout << "k = " << k << endl;
    istringstream sin(input_stream);
    vector<string> result = TopK(k, &sin);
    sort(A.begin(), A.end(), [](const string& a, const string& b) {
      return a.size() < b.size();
    });
    assert(equal(A.end() - k, A.end(), result.begin(), result.end(),
                 [](const string& a, const string& b) {
                   return a.size() == b.size();
                 }));
  }
  return 0;
}

File: static/python/top_k.py
import collections
import itertools
import sys
import random
import heapq


# @include
def top_k(k, stream):
    # Entries are compared by their lengths.
    min_heap = [(len(s), s) for s in itertools.islice(stream, k)]
    heapq.heapify(min_heap)
    for next_string in stream:
        # Push next_string and pop the shortest string in min_heap.
        heapq.heappushpop(min_heap, (len(next_string), next_string))
    return [p[1] for p in heapq.nsmallest(k, min_heap)]
# @exclude


def main():
    for _ in range(1000):
        num = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 10)
        A = [str(random.randint(1, 101)) for _ in range(num)]
        k = random.randint(1, num)
        stream = iter(A)
        result = top_k(k, stream)
        A = sorted(A, key=lambda s: len(s))
        assert collections.Counter(
            [len(a) for a in result]) == collections.Counter(
                [len(a) for a in A[-k:]])


if __name__ == '__main__':
    main()
